include/libtwofloat/arithmetics/double-word-arithmetic.hpp

```diff
@@ -233,5 +233,628 @@ inline two<T> div(const two<T> &x, const two<T> &y) {
+    }
+
+    if(input > local_qd_split_thresh || input < -local_qd_split_thresh) {
+      input *= local_split_factor;
+      temp = local_qd_splitter * input;
+      hi = temp - (temp - input);
+      lo = input - hi;
+      hi *= local_split_factor_2;
+      lo *= local_split_factor_2;
+    } else {
+      temp = local_qd_splitter * input;
+      hi = temp - (temp - input);
+      lo = input - hi;
+    }
+  }
+
+  // Reference: QD / inline.h
+  /* Computes fl(input * input) and err(input * input) */
+  template <typename T>
+  inline T two_sqr(T input, T &err) {
+
+    T twopointzero;
+    if constexpr (std::is_same_v<T, float>) {
+      twopointzero = 2.0f;
+    } else if constexpr (std::is_same_v<T, double>) {
+      twopointzero = 2.0;
+    } else {
+      std::cerr << "Other types not supported\n";
+    }
+
+    T hi, lo;
+    T q = input * input;
+    qd::split(input, hi, lo);
+    err = ((hi * hi - q) + twopointzero * hi * lo) + lo * lo;
+    return q;
+  }
+
+  // Reference QD / inline.h
+  // TODO: make sure this has not been already implemented
+  template <typename T>
+  inline T two_sum(T a, T b, T &err) {
+    T s = a + b;
+    T bb = s - a;
+    err = (a - (s - bb)) - (b - bb);
+    return s;
+  }
+
+} // End namespace qd
+
+namespace dd_real {
+
+  // Reference: QD / dd_inline.h
+  template <typename T>
+  inline two<T> sqr(T input) {
+    T p1, p2;
+    p1 = qd::two_sqr(input, p2);
+    return two<T>{p1, p2};
+  }
+
+  // Reference: QD / dd_inline.h
+  // TODO: make sure this has not been already implemented
+  template <typename T>
+  inline two<T> add(T a, T b) {
+    T s, e;
+    s = qd::two_sum(a, b, e);
+    return two<T> (s, e);
+  }
+
+} // End namespace dd_real
+
+// Reference: QD / dd_inline.h
+/* Round to nearest integer */
+template <typename T>
+inline two<T> nint(const two<T> &input) {
+  T hi = qd::nint(input.h);
+  T lo;
+
+  T zeropointzero, pointfive, onepointzero;
+  if constexpr (std::is_same_v<T, float>) {
+    zeropointzero = 0.0f;
+    pointfive = 0.5f;
+    onepointzero = 1.0f;
+  } else if constexpr (std::is_same_v<T, double>) {
+    zeropointzero = 0.0;
+    pointfive = 0.5;
+    onepointzero = 1.0;
+  } else {
+    std::cerr << "Other types not supported\n";
+  }
+
+  if(hi == input.h) {
+    /* High word is an integer already. Round the low word. */
+    lo = qd::nint(input.l);
+
+    /* Renormalize. This is needed if h = some integer, l = 1/2. */
+    hi = qd::quick_two_sum(hi, lo, lo);
+  }
+  else {
+    /* High word is not an integer */
+    lo = zeropointzero;
+    if(std::abs(hi - input.h) == pointfive && input.l < zeropointzero) {
+      /* There is a tie in the high word, consult the low word
+         to break the tie. */
+      hi -= onepointzero; /* NOTE: This does not cause INEXACT. */
+    }
+  }
+
+  return two<T>{hi, lo};
+}
+
+// Reference: QD / dd_inline.h
+/* Cast to double. */
+template <typename T>
+inline T to_double(const two<T> &input) {
+  return input.h;
```

Comment:
We got the T.eval() function for that. This calculated the sum of h and l first, which is required if the number is not normelized. Numbers may not be normalized if they were calculated by a different arithmetic.

Reply:
Fixed in [3dd021e](https://github.com/esa-tu-darmstadt/twofloat/pull/1/commits/3dd021e300d4b6b1a5d6067f032f4bb5492bb2e9).